src/node/vote_producer.rs

```diff
@@ -0,0 +1,203 @@
+                tracing::error!(target: "bsc::vote", error=%e, "LRU lock poisoned (produced_targets)");
+                return;
+            }
+        }
+    }
+
+    // Enforce warm-up delay since mining started
+    let threshold = warmup_blocks();
+    let cnt = BLOCKS_SINCE_MINING.fetch_add(1, Ordering::Relaxed) + 1;
+    if cnt <= threshold {
+        tracing::trace!(target: "bsc::vote", reason = "warmup", count=cnt, threshold=threshold, "skip vote production");
+        return;
+    }
+    // Clamp to threshold to avoid unbounded growth
+    BLOCKS_SINCE_MINING.store(threshold, Ordering::Relaxed);
+
+    // Parent snapshot is used to determine source (last justified)
+    let parent_number = match target_number.checked_sub(1) {
+        Some(n) => n,
+        None => return,
+    };
+
+    let snap = match snapshot_provider.snapshot(parent_number) {
+        Some(s) => s,
+        None => {
+            tracing::trace!(target: "bsc::vote", reason = "missing-snapshot", parent_number=parent_number, "skip vote production");
+            return;
+        }
+    };
+
+    // Verify local validator is part of current validator set and has a non-zero vote address
+    let my_vote_addr = match bls_signer::global_bls_public_key() {
+        Ok(pk) => pk,
+        Err(_) => return,
+    };
+
+    // O(1) membership via per-snapshot vote address index cached in LRU
+    let guard = VOTEADDR_INDEX.lock();
+    let set = match guard {
+        Ok(mut g) => {
+            if let Some(s) = g.get(&snap.block_hash) {
+                s.clone()
+            } else {
+                let mut hs = HashSet::with_capacity(snap.validators_map.len());
+                for info in snap.validators_map.values() {
+                    if info.vote_addr != VoteAddress::ZERO {
+                        hs.insert(info.vote_addr);
+                    }
+                }
+                g.insert(snap.block_hash, hs.clone());
+                hs
+            }
+        }
+        Err(e) => {
+            tracing::error!(target: "bsc::vote", error=%e, "LRU lock poisoned (voteaddr index)");
+            return;
+        }
+    };
+
+    let is_validator = set.contains(&my_vote_addr);
+    drop(set);
+    if !is_validator {
+        tracing::trace!(target: "bsc::vote", reason = "not-validator", "skip vote production");
+        return;
+    }
+
+    // Too-late-to-vote guard: ensure we have time to broadcast before next block assembly
+    let cur_ms = calculate_millisecond_timestamp(head);
+    let vote_assemble_ms = cur_ms.saturating_add(snap.block_interval);
+    let now_ms = std::time::SystemTime::now()
+        .duration_since(std::time::UNIX_EPOCH)
+        .map(|d| d.as_millis() as u64)
+        .unwrap_or(u64::MAX);
+    if now_ms.saturating_add(TIME_FOR_BROADCAST_MS) > vote_assemble_ms {
+        tracing::trace!(target: "bsc::vote", reason = "too-late", now_ms=now_ms, vote_assemble_ms=vote_assemble_ms, "skip vote production");
+        return;
+    }
+
```

Comment:
pls add a `TODO: add vote journal later to check vote rule, and avoid vote loss due to failure`